### DISCUSSION TITLE
SetPoint Command

### DISCUSSION THREAD

**Discussion by Omikaa** [bitcraze/discussions#1267] [Q&A]
![image](https://github.com/bitcraze/crazyflie-lib-python/assets/136170290/dd1300c3-f20e-4192-ad50-9771053f8e3a)
Hello, I want to know something about the 'set_point' command in the Crazyflie Python API.
Does the value of thrust that is range from 10001 to 60000 mean the motors's PWM? Maybe as far as i can tell, it just means the   value to denote the motors' power.

**Comment by gemenerik** [comment]
They represent the PWM ratio, which can actually take values between 0 and 65535. Below 10k you can expect almost no thrust. [The PWM is limited to 60k in the firmware](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_rpyt.c#L137) to avoid thrust saturation (so that we can still orient the drone correctly when requesting high thrust).

**Comment by gemenerik** [comment]
From #1265
> Is the thrust in command that is range from 10001 to 60000 linear? 

The [PWM is a ratio](https://en.wikipedia.org/wiki/Pulse-width_modulation). In this [example measurement](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/pwm-to-thrust/#software) of pwm to thrust you can see the relation between pwm ratio and thrust is non-linear.

**Comment by Omikaa** [comment]
Now I am training RL task in [pybullet_drones ](https://github.com/utiasDSL/gym-pybullet-drones), but I find that the min_pwm is 20000 and the max_pwm is 65535 in simulation environment.And after training a model to control CF2.1, I scale the range (-1 , 1) to (10000, 60000) which means to map the RL output to the set_point command.However, the CF can fly and hover at 1.5m high in sim.In real world, the output is not enough to give it power to fly to the height.
I am so confused about the pwm in sim and set_point command.Is there any advice to reduce the gap between the sim and real?

**Reply by knmcguire** [reply]
It seems that most likely the motor characteristics of the Crazyflie in pybullet_drones don't match those of your Crazyflie, or perhaps their PWM value actually means something different. Since pybullet_drones is not a simulator we control, there is not much we can help out with that, so you should contact the maintainers of those.

In our our simulation repository we are struggling with this as well. This issue has some overview of some system ID papers we have been looking at as well so perhaps that can pose as reference: https://github.com/bitcraze/crazyflie-simulation/issues/60

**Reply by Omikaa** [reply]
Thanks very much.
